Block scripts with media/csv mimetypes

iamrohit7 opened the pull request:
<!-- Please describe your changes on the following line: -->
I am unsure about setting the `internalResponse` as stated in Step 16 of the `main fetch` standard. I don't see the other algorithms implemented. Should I move this code to a function?

cc @jdm

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #14520  (github issue number if applicable).

<!-- Either: -->
- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because _____

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/14770)
<!-- Reviewable:end -->

Commit:
Block scripts with media/csv mimetypes

highfive commented:
Heads up! This PR modifies the following files:
 * @kichjang: components/net/fetch/methods.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify net code, but no tests are modified. Please consider adding a test!

jdm commented:
I don't think there's any need to create a function for that. However, we should set the `response` variable rather than returning a value, otherwise we miss calling things like `target.process_response()` later on.

asajeffrey commented on components/net/fetch/methods.rs:
```diff
@@ -277,6 +277,19 @@ pub fn main_fetch(request: Rc<Request>,
 
         // Step 16
         // TODO this step (CSP/blocking)
+        if request.type_ == Type::Script {
```
Add a comment that this is handling the "should internalResponse to request be blocked due to its MIME type" case, and add TODO comments for "should internalResponse to request be blocked as mixed content [MIX]", "should internalResponse to request be blocked by Content Security Policy [CSP]" and "should internalResponse to request be blocked due to nosniff"?

asajeffrey commented on components/net/fetch/methods.rs:
```diff
@@ -277,6 +277,19 @@ pub fn main_fetch(request: Rc<Request>,
+                    ContentType(Mime(TopLevel::Video, _, _)) |
+                    ContentType(Mime(TopLevel::Image, _, _)) |
+                    ContentType(Mime(TopLevel::Text, SubLevel::Plain, _)) => {
+                        return Response::network_error(NetworkError::Internal("Invalid repsonse mimetype".into()));
```
We should be setting `response` and `internalResponse` here?

asajeffrey commented on components/net/fetch/methods.rs:
```diff
@@ -277,6 +277,19 @@ pub fn main_fetch(request: Rc<Request>,
+                    ContentType(Mime(TopLevel::Audio, _, _)) |
+                    ContentType(Mime(TopLevel::Video, _, _)) |
+                    ContentType(Mime(TopLevel::Image, _, _)) |
+                    ContentType(Mime(TopLevel::Text, SubLevel::Plain, _)) => {
```
Should this be `text/csv` rather than `text/plain`?

iamrohit7 commented:
@asajeffrey I addressed all your comments in the commits. I had to mix this part of Step 16 with Step 14 because the borrow checker wouldn't agree to compile the code.

KiChjang commented:
This should really wait for #14620 to be merged first, since some of the questions raised here in this PR is addressed in that PR.

iamrohit7 commented:
@KiChjang That would make this more easy :) I had to solve the same borrow checker problem @dpyro solved, but his solution is much simpler

asajeffrey commented on components/net/fetch/methods.rs:
```diff
@@ -271,22 +271,43 @@ pub fn main_fetch(request: Rc<Request>,
         let network_error_res;
-        let internal_response = if let Some(error) = response.get_network_error() {
+        let mut step16 = false;
+        let internal_response = if invalid_mime && request.type_ == Type::Script {
```
Would it be easier to follow the spec here if we declared `internal_response` to be mutable, then updated it as per steps 15-18?

iamrohit7 replied:
No, because `internal_response` is a reference to `response`, they can't be changed at the same time due to issues with borrowing. So, `internal_response` had to be set at once as a result of step 14 and step 16.

asajeffrey replied:
Can we not `clone()` it when we discover we need to mutate `internal_response`?

iamrohit7 replied:
@KiChjang suggested against it in #14620 as it is expensive for a Response.

asajeffrey commented:
@iamrohit7 I added the "blocked on external" label while we're waiting on #14620.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #14620) made this pull request unmergeable. Please resolve the merge conflicts.

jdm commented:
No longer blocked! Hooray!

nox commented:
Oops, author left Github, need to salvage this.
